### PR TITLE
update PreferenceSettings to include multiple locations for packages

### PR DESCRIPTION
**Pull request by nguyen-binh-minh**
@Benglin PTAL

### CODE REVIEW

File: src/DynamoCore/Core/PathManager.cs [hunk 1]
```diff
@@ -355,6 +355,17 @@ internal string GetBackupFilePath(WorkspaceModel workspace)
             return Path.Combine(BackupDirectory, fileName);
         }
 
+        internal void LoadPackageFolders(IEnumerable<string> folders)
+        {
+            // Preserve the priority specified in 'folders' list by first removing
+            // elements in 'packageDirectories' that are also present in 'folders'
+            var remainingFolders = packageDirectories.Where(x => !folders.Contains(x));
+            packageDirectories.Clear();
```
Comment (Benglin) [inline src/DynamoCore/Core/PathManager.cs]: Due to lazy evaluation, `remainingFolders` will be empty as soon as `packageDirectories.Clear` is called. So it's always empty moving forward.

**Comment by jnealb** [comment]
@nguyen-binh-minh @Benglin We need to use the submission templates, they seem to be missing? Thanks for clarifying.

File: src/DynamoCore/Core/PreferenceSettings.cs [hunk 2]
```diff
@@ -102,6 +102,16 @@ public int MaxNumRecentFiles
...
+        /// A list of folders containing custom nodes.
+        /// </summary>
+        public List<string> CustomNodeFolders { get; set; }
+
+        /// <summary>
+        /// A list of folders containing zero-touch nodes.
+        /// </summary>
+        public List<string> PackageFolders { get; set; } 
+
+        /// <summary>
```
Comment (Benglin) [inline src/DynamoCore/Core/PreferenceSettings.cs]: These should be returned as IEnumerable instead.

File: src/DynamoCore/Interfaces/IPathResolver.cs [hunk 3]
```diff
@@ -70,9 +70,15 @@ public interface IPathManager
...
+        /// Directories from where custom nodes are to be loaded. The implementor
+        /// of this interface method should always guarantee that a non-empty 
+        /// list is returned, and that the first entry represents the default 
+        /// custom node directory. Custom nodes created are stored in the
+        /// default directory, which is specific to the current user. Changes to
+        /// custom nodes may or may not be saved to their current location depeding
+        /// on write access.
         /// </summary>
-        string UserDefinitions { get; }
+        IEnumerable<string> UserDefinitionDirectories { get; }
```
Comment (Benglin) [inline src/DynamoCore/Interfaces/IPathResolver.cs]: Please check with Ritesh to see if this is needed. We will not have separated folders for `packages` and `definitions`. All under these custom folders are just loose dyf and package folders like `Mantis`

File: src/DynamoCore/Core/PreferenceSettings.cs (hunk 2 above)
Comment (nguyen-binh-minh) [inline src/DynamoCore/Core/PreferenceSettings.cs]: If they are IEnumerable, serialization will fail. Why should they be IEnumerable while BackupFiles is List<string>?

**Comment by Benglin** [comment]
Thanks for looking, @jnealb . This pr is for my own repo, there's no need to use the template as it will be used when we merge the final work into master.

File: src/DynamoCore/Core/PreferenceSettings.cs (hunk 2 above)
Comment (Benglin) [inline src/DynamoCore/Core/PreferenceSettings.cs]: Ah, forgot about serialization! Then list makes sense

File: src/DynamoCore/Core/PreferenceSettings.cs [hunk 4]
```diff
@@ -260,7 +273,25 @@ public static PreferenceSettings Load(string filePath)
...
+            foreach (var folder in folders)
+            {
+                if (!lookup.Contains(folder))
+                {
+                    lookup.Add(folder);
+                    uniqueFolders.Add(folder);
+                }
+            }
+            return uniqueFolders;
+        }
```
Comment (Benglin) [inline src/DynamoCore/Core/PreferenceSettings.cs]: Can we not use `List.Unique`?

File: src/DynamoPackages/PackageLoader.cs [hunk 5]
```diff
@@ -178,6 +179,18 @@ private void ScanAllPackageDirectories(IPreferences preferences)
...
+                rootDirectoryExists = Directory.Exists(root);
+            });
+
+            t.Start();
+            bool requestCompleted = t.Join(5000);
+
+            if (!requestCompleted || !rootDirectoryExists)
+                return;
+
             foreach (var dir in
```
Comment (Benglin) [inline src/DynamoPackages/PackageLoader.cs]: I don't think I understand the intent of using thread here...

**Comment by jnealb** [comment]
@Benglin np

**Pull request merged**